File: set_finder/core/src/card_extractor/image_filter.cpp

#include "image_filter.h"

#include <opencv2/core/types.hpp>
#include <opencv2/imgproc.hpp>

#include "utils/logger.h"

namespace
{
constexpr auto SIGMA_X = 0;
}

ImageFilter::ImageFilter(const Image& source_img)
{
    const auto kernel = cv::Size(7, 7);

    // It works faster with blur
    cv::GaussianBlur(source_img, m_blured_img, kernel, SIGMA_X);
    cv::cvtColor(m_blured_img, m_hsv_img, cv::COLOR_BGR2HSV);

    const auto nopixels = static_cast<int>(m_hsv_img.size().area() * 0.20);

    std::vector<cv::Mat> channels(3);
    cv::split(m_hsv_img, channels);
    auto s_channel = channels.at(1);
    auto v_channel = channels.at(2);

    const auto flatten_s_channel = s_channel.reshape(1, 1);
    cv::Mat s_sorted{};
    cv::sort(flatten_s_channel, s_sorted, cv::SORT_ASCENDING | cv::SORT_EVERY_ROW);
    m_s_mean = static_cast<int>(cv::mean(s_sorted.colRange(0, nopixels))[0]);

    const auto flatten_v_channel = v_channel.reshape(1, 1);
    cv::Mat v_sorted{};
    cv::sort(flatten_v_channel, v_sorted, cv::SORT_DESCENDING | cv::SORT_EVERY_ROW);
    m_v_mean = static_cast<int>(cv::mean(v_sorted.colRange(0, nopixels))[0]);
}

Image ImageFilter::filter(const Boundary& lower, const Boundary& upper) const
{
    Image mask{}, filtered_img{};

    cv::inRange(m_hsv_img, lower, upper, mask);
    cv::bitwise_and(m_blured_img, m_blured_img, filtered_img, mask);

    return filtered_img;
}

Image ImageFilter::filter(const Boundaries& boundaries) const
{
    return filter(boundaries.first, boundaries.second);
}

Image ImageFilter::get_hsv_img() const
{
    return m_hsv_img;
}

int ImageFilter::get_s_mean() const
{
    return m_s_mean;
}

int ImageFilter::get_v_mean() const
{
    return m_v_mean;
}

File: set_finder/poc/set_finder/image_filter.py
import cv2
import numpy as np


class ImageFilter:
    def __init__(self, sourceImg):
        sourceImg = cv2.GaussianBlur(sourceImg, (7, 7), 0)
        self._sourceImg = sourceImg
        self._hsvImg = cv2.cvtColor(sourceImg, cv2.COLOR_BGR2HSV)

        nopixels = int(self._hsvImg.shape[0] * self._hsvImg.shape[1] * 0.20)
        sChannel = self._hsvImg[:, :, 1]
        self._sMean = int(cv2.mean(np.sort(sChannel.flatten())[:nopixels])[0])

        vChannel = self._hsvImg[:, :, 2]
        self._vMean = int(cv2.mean(np.sort(vChannel.flatten())[:nopixels])[0])

        sHistr = cv2.calcHist([self._hsvImg], [1], None, [256], [0, 256])
        self._sPeak = np.argmax(sHistr)

        vHistr = cv2.calcHist([self._hsvImg], [2], None, [256], [0, 256])
        self._vPeak = np.argmax(vHistr)

    def getSMean(self):
        return self._sMean

    def filter(self, lower, upper):
        mask = cv2.inRange(self._hsvImg, lower, upper)
        return cv2.bitwise_and(self._sourceImg, self._sourceImg, mask=mask)

    def adaptiveColorDetectorFilter(self, lower, upper):
        sMinOffset = self._sPeak + 10

        adaptedUpper = upper.copy()
        adaptedLower = lower.copy()

        adaptedLower[1] += sMinOffset

        mask = cv2.inRange(self._hsvImg, adaptedLower, adaptedUpper)
        return cv2.bitwise_and(self._sourceImg, self._sourceImg, mask=mask)
